packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -432,34 +432,66 @@ export class FluidDataStoreRuntime
 		return context.getChannel();
 	}
 
-	public createChannel(id: string = uuid(), type: string): IChannel {
+	public addChannel(channel: IChannel): void {
+		const id = channel.id;
 		if (id.includes("/")) {
 			throw new UsageError(`Id cannot contain slashes: ${id}`);
 		}
 
 		this.verifyNotClosed();
 
+		assert(!this.contexts.has(id), "addChannel() with existing ID");
+
+		const type = channel.attributes.type;
+		const factory = this.sharedObjectRegistry.get(channel.attributes.type);
+		if (factory === undefined) {
+			throw new Error(`Channel Factory ${type} not registered`);
+		}
+
+		this.addOrCreateChannelCore(channel);
+	}
+
+	public createChannel(id: string = uuid(), type: string): IChannel {
+		if (id.includes("/")) {
+			throw new UsageError(`Id cannot contain slashes: ${id}`);
+		}
+
+		this.verifyNotClosed();
 		assert(!this.contexts.has(id), 0x179 /* "createChannel() with existing ID" */);
-		this.notBoundedChannelContextSet.add(id);
+
+		assert(type !== undefined, 0x209 /* "Factory Type should be defined" */);
+		const factory = this.sharedObjectRegistry.get(type);
+		if (factory === undefined) {
+			throw new Error(`Channel Factory ${type} not registered`);
+		}
+
+		const channel = factory.create(this, id);
+		this.addOrCreateChannelCore(channel);
+
+		return channel;
+	}
+
+	private addOrCreateChannelCore(channel: IChannel) {
```

Comment:
I think this probably reads better if this remains part of `addChannel` and `createChannel` just calls `addChannel` after creating the channel.  So just 2 methods.

Rather than splitting out this "core" method (which is really more like `addChannelCore` since it's not actually creating any channels).

Reply:
core just means it is the core or common code for both of those methods. Looking at the code, I don't think createChannel can call addChannel based on what they are returning and the validation they are doing.

Reply:
I suppose it will depend on the checks conversation above, but probably at least rename then as a nit:

```suggestion
	private addChannelCore(channel: IChannel) {
```

Reply:
Renamed it to createChannelContext()